antithesis_sdk.h

```diff
@@ -4,8 +4,28 @@
 //
 // Documentation for the SDK is found at https://antithesis.com/docs/using_antithesis/sdk/cpp/overview/.
 
+#if __cplusplus < 202000L
+    #error "The Antithesis C++ API requires C++20 or higher"
+    #ifndef NO_ANTITHESIS_SDK
```

Comment:
I do not understand the point of this; if `#error` was encountered the program is ill-formed and compilation stops https://en.cppreference.com/w/c/preprocessor/error , so what's the point of doing anything below `#error` ?

If anything, I would rather perform this check the other way round:

```
#ifndef NO_ANTITHESIS_SDK
  #if __cplusplus < 202000L
    #error "The Antithesis C++ API requires C++20 or higher"
  #endif
  #if !defined(__clang__)
    #error "The Antithesis C++ API requires clang compiler"
  #endif
  #if __clang_major__ < 16
    #error "The Antithesis C++ API requires clang version 16 or higher"
  #endif
#endif
```

Reply:
Hmmmm, clang doesn't seem to be following the standard then.

This code:
```c++
#if __cplusplus < 202000L
    #error "The Antithesis C++ API requires C++20 or higher"
    #ifndef NO_ANTITHESIS_SDK
        #define NO_ANTITHESIS_SDK
    #endif
#endif

THIS IS INVALID CODE
```

Generates multiple errors if you compile with C++17 (`clang++ --std=c++17 foo.cpp`). First the error "The Antithesis C++ API requires C++20 or higher" then a bunch (10 or so lines) related to the line THIS IS INVALID CODE.

The notion of the `#error`s and the various `#ifdefs` was that rather than getting 20 or 30 lines of really cryptic template-heavy messages, you'd get a single message telling you the root cause. Without the `#error`, you'd tend to get weird messages where you're not sure if it's your code that's the problem or ours or compiler version or what. Which we felt would be incredibly confusing.

I **do** 100% agree that we should wrap the checks in `#ifndef NO_ANTITHESIS_SDK` - if you're compiling with NO_ANTITHESIS_SDK, this file should be a no-op rather than causing your compile to fail. That's a great catch.

Reply:
Code can be ill-formed for more than one reason :) Good catch though, I did not think of it.

Reply:
Thanks for both of your suggestions! The error detection and reporting is indeed better this way!